scriptworker/task.py

```diff
@@ -422,35 +391,59 @@ def is_try(task, source_env_prefix):
     ))
 
 
-def is_pull_request(context, task):
+async def is_pull_request(context, task):
     """Determine if a task is a pull-request-like task (restricted privs).
 
-    This goes further than checking ``tasks_for``.  We may or may not want
+    This goes further than checking ``tasks_for``. We may or may not want
     to keep this.
 
     This checks for the following things::
 
         * ``task.extra.env.tasks_for`` == "github-pull-request"
         * ``task.payload.env.MOBILE_HEAD_REPOSITORY`` doesn't come from an official repo
         * ``task.metadata.source`` doesn't come from an official repo, either
+        * The last 2 items are landed on the official repo
 
     Args:
         context (scriptworker.context.Context): the scriptworker context.
         task (dict): the task definition to check.
 
     Returns:
-        bool: True if it's a pull-request
+        bool: True if it's a pull-request. False if it either comes from the official repos or if
+        the origin can't be determined. In fact, some valid scriptworker tasks don't
 
     """
-    repo = get_repo(task, context.config['source_env_prefix']) or ''
-    return any((
-        task.get('extra', {}).get('tasks_for') == 'github-pull-request',
-        not _does_url_come_from_official_repo(context, repo),
-        not _does_url_come_from_official_repo(context, task['metadata'].get('source', '')),
-    ))
+    tasks_for = task.get('extra', {}).get('tasks_for')
+    repo_url_from_payload = get_repo(task, context.config['source_env_prefix'])
+    revision_from_payload = get_revision(task, context.config['source_env_prefix'])
+
+    metadata_source_url = task['metadata'].get('source', '')
+    repo_from_source_url, revision_from_source_url = \
+        extract_github_repo_and_revision_from_source_url(metadata_source_url)
+
+    conditions = [tasks_for == 'github-pull-request']
+    urls_revisions_and_can_skip = (
+        (repo_url_from_payload, revision_from_payload, True),
+        (repo_from_source_url, revision_from_source_url, False),
+    )
+    for repo_url, revision, can_skip in urls_revisions_and_can_skip:
+        # XXX In the case of scriptworker tasks, neither the repo nor the revision is defined
+        if not repo_url and can_skip:
```

Comment:
What happens if `not repo_url` and `can_skip is False`? We expect `extract_github_repo_and_name(...)` to throw an error?

Reply:
That's right. `repo_url` won't be a valid github URL and then `_check_github_url_is_supported()` raises a `ValueError`